### PR TITLE
gh-124210: Add introduction to `threading` docs

### PR DESCRIPTION
**Pull request by donbarbos**
I chose to take courage and add not only a usage example but also an introduction to the documentation.
I was inspired by the documentation at https://docs.python.org/3/library/multiprocessing.html
This page also have introduction before basic example

* Issue: gh-124210

----
📚 Documentation preview 📚: https://cpython-previews--127046.org.readthedocs.build/en/127046/library/threading.html#introduction

### CODE REVIEW

**Comment by donBarbos** [comment]
@merwok could you check the changes of this PR?

File: Doc/library/threading.rst [hunk 1]
```diff
@@ -8,6 +8,54 @@
...
 --------------
 
+Introduction
+------------
+
+The :mod:`threading` module provides a way to run multiple threads (smaller
+units of a process) concurrently within a single process. It allows for the
+creation and management of threads, making it possible to execute tasks in
+parallel, sharing memory space. Threads are particularly useful when tasks are
+I/O-bound, such as reading from or writing to files, or making network requests,
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: This caught me off guard. Typically, threads are for speeding up CPU bound operations, and something like `asyncio` is for I/O bound tasks. Threads being more efficient for I/O bound tasks is really only because of the GIL, but we're trying to get rid of that anyway.

File: Doc/library/threading.rst [hunk 2]
```diff
@@ -8,6 +8,54 @@
...
+where much of the time is spent waiting for external resources.
+
+Unlike the :mod:`multiprocessing` module, which uses separate processes to
+bypass the :term:`Global Interpreter Lock <global interpreter lock>`, the
+threading module operates within a single process, meaning that all threads
+share the same memory space. However, the GIL limits the performance gains of
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]:
```suggestion
share the same memory space. However, the :term:`GIL` limits the performance gains of
```

File: Doc/library/threading.rst [hunk 3]
```diff
@@ -8,6 +8,54 @@
...
+threading when it comes to CPU-bound tasks, as only one thread can execute
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: Ah, here we are about CPU bound tasks. I think we should mention the experimental free-threaded builds here.

File: Doc/library/threading.rst [hunk 4]
```diff
@@ -8,6 +8,54 @@
...
+Python bytecode at a time. Despite this, threading remains a useful tool for
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]:
```suggestion
Python bytecode at a time. Despite this, threads remain a useful tool for
```

File: Doc/library/threading.rst (hunk 2 above)
Comment (ZeroIntensity) [reply Doc/library/threading.rst]: We should also mention that this is specific to CPython.

File: Doc/library/threading.rst [hunk 5]
```diff
@@ -8,6 +8,54 @@
...
+       # Using `args` to pass positional arguments and `kwargs` for keyword arguments
+       t = threading.Thread(target=crawl, args=(link,), kwargs={"delay": 2})
+       threads.append(t)
+       t.start()
+
+   # Wait for all threads to finish
+   for t in threads:
+       t.join()
+
 This module constructs higher-level threading interfaces on top of the lower
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: I think it looks kind of weird having all this technical information below the introduction. This, and the parts below it down to the note about WASI compatibility, should ideally go before the introduction.

**Review by ZeroIntensity — COMMENTED** [review]
Sorry about the delay on reviews!

File: Doc/library/threading.rst (hunk 5 above)
Comment (donbarbos) [reply Doc/library/threading.rst]: I looked at `multiprocessing.rst` as a reference and it's in the same order:
```rst
Introduction
------------
Common info about module

   python code example
```

or you mean information about `This module constructs higher-level threading interfaces ...` needs to be moved before `Introduction`?

Comment (ZeroIntensity) [reply Doc/library/threading.rst]: > or you mean information about This module constructs higher-level threading interfaces ... needs to be moved before Introduction?

Yeah, as well as the stuff about compatibility, `versionadded`, and whatnot.

Comment (donbarbos) [reply Doc/library/threading.rst]: like this?
📚 Documentation preview 📚: https://cpython-previews--127046.org.readthedocs.build/en/127046/library/threading.html#introduction

File: Doc/library/threading.rst [hunk 6]
```diff
@@ -6,8 +6,6 @@
 
 **Source code:** :source:`Lib/threading.py`
 
---------------
-
```
Comment (merwok) [inline Doc/library/threading.rst]: This was added on purpose, in all docs that use the `Source code:` line. Please restore, and in general be more cautious about removing things 🙂

Comment (donbarbos) [reply Doc/library/threading.rst]: it's not removed but moved, it's still there, above Introduction.
should I add one or move the existing one?

Comment (merwok) [reply Doc/library/threading.rst]: Keep it here!

Comment (donbarbos) [reply Doc/library/threading.rst]: :+1: moved it back

File: Doc/library/threading.rst (hunk 3 above)
Comment (donbarbos) [reply Doc/library/threading.rst]: I added mention

File: Doc/library/threading.rst [hunk 7]
```diff
@@ -47,6 +47,57 @@ level :mod:`_thread` module.
...
+Unlike the :mod:`multiprocessing` module, which uses separate processes to
+bypass the :term:`Global Interpreter Lock <global interpreter lock>`, the
+threading module operates within a single process, meaning that all threads
+share the same memory space. However, the :term:`GIL` limits the performance gains of
+threading when it comes to CPU-bound tasks, as only one thread can execute
+Python bytecode at a time. Despite this, threads remain a useful tool for
+achieving concurrency in many scenarios.
+
+*(Experimental free-threaded builds of CPython disable the GIL, enabling true
+parallel execution of threads, but this feature is still under development.)*
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: - I don't think this should be italicized. 
- The phrase "experimental" already implies "under development."
- We should use some terms and link PEP-703.
```suggestion
As of Python 3.13, experimental :term:`free-threaded <free threading>` builds can disable the GIL, enabling true parallel execution of threads, but this feature is not available by default. (See :pep:`703`.)
```

File: Doc/library/threading.rst [hunk 8]
```diff
@@ -47,6 +47,57 @@ level :mod:`_thread` module.
...
+A typical use case for :mod:`threading` includes managing a pool of worker
+threads that can process multiple tasks concurrently.  This basic example of
+creating and starting threads using :class:`~threading.Thread`::
+
+   import threading
+   import time
+
+   def crawl(link, delay=3):
+       print(f"crawl started for {link}")
+       time.sleep(delay)
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: A comment to clarify would be good:
```suggestion
       time.sleep(delay)  # Blocking I/O, simulating a network request
```

File: Doc/library/threading.rst [hunk 9]
```diff
@@ -47,6 +47,57 @@ level :mod:`_thread` module.
...
+       print(f"crawl ended for {link}")
+
+   links = [
+       "https://example.com",
+       "https://another-example.com",
+       "https://yet-another-example.com"
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: I always like example URLs to be real. I'll just suggest using Python ones for now, but feel free to have some fun with them.
```suggestion
       "https://python.org",
       "https://docs.python.org",
       "https://peps.python.org"
```

File: Doc/library/threading.rst [hunk 10]
```diff
@@ -47,6 +47,57 @@ level :mod:`_thread` module.
...
+   ]
+
+   # Start threads for each link
+   threads = []
+   for link in links:
+       # Using `args` to pass positional arguments and `kwargs` for keyword arguments
+       t = threading.Thread(target=crawl, args=(link,), kwargs={"delay": 2})
+       threads.append(t)
+       t.start()
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: It's generally more common to want to start all threads at once, rather than incrementally as `Thread` objects are created. I'd arrange it into a new loop, but I'm not married to this suggestion:

```suggestion

    # Start each thread
    for t in threads:
       t.start()
```

File: Doc/library/threading.rst [hunk 11]
```diff
@@ -47,6 +47,57 @@ level :mod:`_thread` module.
...
+I/O-bound, such as reading from or writing to files, or making network requests,
+where much of the time is spent waiting for external resources.
+
+Unlike the :mod:`multiprocessing` module, which uses separate processes to
+bypass the :term:`Global Interpreter Lock <global interpreter lock>`, the
+threading module operates within a single process, meaning that all threads
+share the same memory space. However, the :term:`GIL` limits the performance gains of
+threading when it comes to CPU-bound tasks, as only one thread can execute
+Python bytecode at a time. Despite this, threads remain a useful tool for
+achieving concurrency in many scenarios.
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: Note that the GIL is a CPython implementation detail. We *could* just add a "this only applies to CPython" phrase, but that might make things too hard to follow. I'd suggest splitting this into two sections:

- One section titled "Introduction" as you already have it, containing the first paragraph and then the example below this paragraph.
- A second section below entitled "The GIL" (or something along those lines), and then move this paragraph to that section.

File: Doc/library/threading.rst [hunk 12]
```diff
@@ -47,6 +47,57 @@ level :mod:`_thread` module.
 
 .. include:: ../includes/wasm-notavail.rst
 
+Introduction
+------------
+
+The :mod:`threading` module provides a way to run multiple threads (smaller
+units of a process) concurrently within a single process. It allows for the
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: Instead of trying to define threads in a tiny parenthesis, it *might* be worth just adding a hyperlink to the Wikipedia page for threads or something like that. It's fine to keep it as-is if you want, though.

File: Doc/library/threading.rst [hunk 13]
```diff
@@ -47,6 +47,57 @@ level :mod:`_thread` module.
...
+creation and management of threads, making it possible to execute tasks in
+parallel, sharing memory space. Threads are particularly useful when tasks are
+I/O-bound, such as reading from or writing to files, or making network requests,
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: This comma hurts clarity, so let's just shorten it:
```suggestion
I/O-bound, such as file operations or making network requests,
```

File: Doc/library/threading.rst (hunk 11 above)
Comment (donbarbos) [reply Doc/library/threading.rst]: I called second section "GIL and Performance Considerations"

File: Doc/library/threading.rst [hunk 14]
```diff
@@ -47,6 +47,64 @@ level :mod:`_thread` module.
 
 .. include:: ../includes/wasm-notavail.rst
 
+Introduction
+------------
+
+The :mod:`threading` module provides a way to run multiple threads (smaller
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: Thinking out loud: maybe we want to link to a Wikipedia page (or some other external reference) on threads? That's probably better to someone clueless rather than "smaller units of a process."

File: Doc/library/threading.rst [hunk 15]
```diff
@@ -47,6 +47,64 @@ level :mod:`_thread` module.
...
+units of a process) concurrently within a single process. It allows for the
+creation and management of threads, making it possible to execute tasks in
+parallel, sharing memory space. Threads are particularly useful when tasks are
+I/O-bound, such as file operations or making network requests,
+where much of the time is spent waiting for external resources.
+
+A typical use case for :mod:`threading` includes managing a pool of worker
+threads that can process multiple tasks concurrently.  This basic example of
```
Comment (ZeroIntensity) [inline Doc/library/threading.rst]: "this" makes it a dependent clause. It should be independent:
```suggestion
threads that can process multiple tasks concurrently.  Here's a basic example of
```

**Review by ZeroIntensity — COMMENTED** [review]
Sorry for the delay. This looks just about ready.

File: Doc/library/threading.rst (hunk 14 above)
Comment (donbarbos) [reply Doc/library/threading.rst]: I added link to [wikipedia page](https://en.wikipedia.org/wiki/Thread_(computing))

**Review by ZeroIntensity — APPROVED** [review]
Thanks, LGTM.

**Comment by donBarbos** [comment]
@merwok what do you think about this?

File: Doc/library/threading.rst [hunk 16]
```diff
@@ -47,6 +47,65 @@ level :mod:`_thread` module.
...
+share the same memory space. However, the :term:`GIL` limits the performance
+gains of threading when it comes to CPU-bound tasks, as only one thread can
+execute Python bytecode at a time. Despite this, threads remain a useful tool
+for achieving concurrency in many scenarios.
+
+As of Python 3.13, experimental :term:`free-threaded <free threading>` builds
+can disable the :term:`GIL`, enabling true parallel execution of threads, but
+this feature is not available by default. (See :pep:`703`.)
+
 This module defines the following functions:
```
Comment (merwok) [inline Doc/library/threading.rst:114]: I think a heading should be added before starting this section. At the moment we’re still in the GIL section!

File: Doc/library/threading.rst [hunk 17]
```diff
@@ -47,6 +47,65 @@ level :mod:`_thread` module.
 
 .. include:: ../includes/wasm-notavail.rst
 
+Introduction
+------------
```
Comment (merwok) [inline Doc/library/threading.rst:17]: What do you think of writing this intro _before_ the pile of directives above?

Comment (donbarbos) [reply Doc/library/threading.rst:17]: as far as I know accessibility directives should be at the top so i don't think we can move `Introduction` higher

File: Doc/library/threading.rst (hunk 16 above)
Comment (donbarbos) [reply Doc/library/threading.rst:114]: thank you, but it's hard for me to find a suitable heading with my poor English.
what about adding `Reference` header like for [multiprocessing doc](https://docs.python.org/dev/library/multiprocessing.html#reference)
and make the following headings subheadings of this part?

File: Doc/library/threading.rst (hunk 17 above)
Comment (merwok) [reply Doc/library/threading.rst:17]: But here we have a change note, see also, naming note, implementation details which should all be after the simple intro.

Comment (donbarbos) [reply Doc/library/threading.rst:17]: done. I like that now the implementation details about GIL are located next to the section about GIL

File: Doc/library/threading.rst [hunk 18]
```diff
@@ -45,7 +91,23 @@ level :mod:`_thread` module.
    However, threading is still an appropriate model if you want to run
    multiple I/O-bound tasks simultaneously.
 
-.. include:: ../includes/wasm-notavail.rst
+GIL and Performance Considerations
```
Comment (hugovk) [inline Doc/library/threading.rst]:
```suggestion
GIL and performance considerations
```
And we have a mix of headers on this page, I suggest updating the others to use sentence case as well.

https://devguide.python.org/documentation/style-guide/#capitalization

File: Doc/library/threading.rst [hunk 19]
```diff
@@ -11,6 +11,52 @@
...
+
+Introduction
+------------
+
+The :mod:`threading` module provides a way to run multiple `threads
+<https://en.wikipedia.org/wiki/Thread_(computing)>`_ (smaller
+units of a process) concurrently within a single process. It allows for the
+creation and management of threads, making it possible to execute tasks in
+parallel, sharing memory space. Threads are particularly useful when tasks are
+I/O-bound, such as file operations or making network requests,
```
Comment (hugovk) [inline Doc/library/threading.rst]:
```suggestion
I/O bound, such as file operations or making network requests,
```

File: Doc/library/threading.rst [hunk 20]
```diff
@@ -11,6 +11,52 @@
 This module constructs higher-level threading interfaces on top of the lower
 level :mod:`_thread` module.
 
+.. include:: ../includes/wasm-notavail.rst
+
+Introduction
+------------
+
+The :mod:`threading` module provides a way to run multiple `threads
```
Comment (hugovk) [inline Doc/library/threading.rst]:
```suggestion
The :mod:`!threading` module provides a way to run multiple `threads
```

> Do not use a link when the link would point to the current unit. It’s natural to use the name of a function in the documentation for the function, but a link on that function name that simply reloads the section the user is already reading is useless and distracting.

https://devguide.python.org/documentation/style-guide/#links

File: Doc/library/threading.rst [hunk 21]
```diff
@@ -11,6 +11,52 @@
...
+<https://en.wikipedia.org/wiki/Thread_(computing)>`_ (smaller
+units of a process) concurrently within a single process. It allows for the
+creation and management of threads, making it possible to execute tasks in
+parallel, sharing memory space. Threads are particularly useful when tasks are
+I/O-bound, such as file operations or making network requests,
+where much of the time is spent waiting for external resources.
+
+A typical use case for :mod:`threading` includes managing a pool of worker
```
Comment (hugovk) [inline Doc/library/threading.rst]:
```suggestion
A typical use case for :mod:`!threading` includes managing a pool of worker
```

File: Doc/library/threading.rst [hunk 22]
```diff
@@ -45,7 +91,23 @@ level :mod:`_thread` module.
    However, threading is still an appropriate model if you want to run
    multiple I/O-bound tasks simultaneously.
 
-.. include:: ../includes/wasm-notavail.rst
+GIL and Performance Considerations
+----------------------------------
+
+Unlike the :mod:`multiprocessing` module, which uses separate processes to
+bypass the :term:`Global Interpreter Lock <global interpreter lock>`, the
```
Comment (hugovk) [inline Doc/library/threading.rst]:
```suggestion
bypass the :term:`global interpreter lock` (GIL), the
```

File: Doc/library/threading.rst [hunk 23]
```diff
@@ -45,7 +91,23 @@ level :mod:`_thread` module.
...
+threading module operates within a single process, meaning that all threads
+share the same memory space. However, the :term:`GIL` limits the performance
```
Comment (hugovk) [inline Doc/library/threading.rst]:
```suggestion
share the same memory space. However, the GIL limits the performance
```

File: Doc/library/threading.rst [hunk 24]
```diff
@@ -45,7 +91,23 @@ level :mod:`_thread` module.
...
+gains of threading when it comes to CPU-bound tasks, as only one thread can
+execute Python bytecode at a time. Despite this, threads remain a useful tool
+for achieving concurrency in many scenarios.
+
+As of Python 3.13, experimental :term:`free-threaded <free threading>` builds
+can disable the :term:`GIL`, enabling true parallel execution of threads, but
```
Comment (hugovk) [inline Doc/library/threading.rst]:
```suggestion
can disable the GIL, enabling true parallel execution of threads, but
```

File: Doc/library/threading.rst [hunk 25]
```diff
@@ -45,7 +91,23 @@ level :mod:`_thread` module.
...
+this feature is not available by default. (See :pep:`703`.)
```
Comment (hugovk) [inline Doc/library/threading.rst]:
```suggestion
this feature is not available by default (see :pep:`703`).
```

---

At some point this will become default (possibly in [3.14](https://peps.python.org/pep-0779/)?) so whenever that occurs we'll have to try and remember to fix it. Or wait for the bug report :)

Comment (donbarbos) [reply Doc/library/threading.rst]: I added comment:
```rst
.. TODO: At some point this feature will become available by default.
```
I think when someone changes the default build they might go to this document and stumble upon a comment so there is a higher chance that they will not forget to correct this paragraph

Comment (ZeroIntensity) [reply Doc/library/threading.rst]: > At some point this will become default (possibly in [3.14](https://peps.python.org/pep-0779/)?)

Woah, what? I thought PEP 779 was supposed to get rid of the "experimental" in "experimental free-threading build," not make it the default.

Comment (hugovk) [reply Doc/library/threading.rst]: Of course, that's right, the default would be a later release 😅

**Review by hugovk — APPROVED** [review]
